keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/DeliveryConfigController.kt

```diff
@@ -60,16 +67,37 @@ class DeliveryConfigController(
   fun upsert(
     @RequestBody
     @SwaggerRequestBody(
-      description = "The delivery config. If its `name` matches an existing delivery config the operation is an update, otherwise a new delivery config is created."
+      description = "The delivery config. If its `name` matches an existing delivery config the operation is " +
+        "an update, otherwise a new delivery config is created."
     )
     deliveryConfig: SubmittedDeliveryConfig
-  ): DeliveryConfig =
-    deliveryConfigProcessors.applyAll(deliveryConfig)
+  ): DeliveryConfig {
+    val metadata: MutableMap<String,Any?> = deliveryConfig.metadata?.toMutableMap() ?: mutableMapOf()
+    val gitMetadata: GitMetadata? = try {
+      val candidateMetadata = metadata.getOrDefault("gitMetadata", null)
+      if (candidateMetadata != null) {
+        mapper.convertValue<GitMetadata>(candidateMetadata)
+      } else {
+        null
+      }
+    } catch (e: IllegalArgumentException) {
+      log.debug("Error converting git metadata ${metadata.getOrDefault("gitMetadata", null)}: {}", e)
+      // not properly formed, so ignore the metadata and move on
+      null
+    }
+
+    // we don't save this key to the database, since we're programmatically adding it in orca.
+    metadata.remove("gitMetadata")
+
+    deliveryConfigProcessors.applyAll(deliveryConfig.copy(metadata = metadata))
       .let { processedDeliveryConfig ->
         validator.validate(processedDeliveryConfig)
         log.debug("Upserting delivery config '${processedDeliveryConfig.name}' for app '${processedDeliveryConfig.application}'")
-        repository.upsertDeliveryConfig(processedDeliveryConfig)
+        val config = repository.upsertDeliveryConfig(processedDeliveryConfig)
+        publisher.publishEvent(DeliveryConfigChangedNotification(config, gitMetadata))
```

Comment:
nit: Should this be encapsulated in a `DeliveryConfigProcessor` like other similar processing we already do?

Reply:
I don't think so, since we're not modifying the delivery config or changing values. We're just gathering information to send a notification. I want to also add validation notifications here (so we can notify with a good message if there's a validation error). Maybe when I play with that it'll make more sense to refactor this area.